### PR TITLE
Fix dislocated parent tasks in FactoryBot - Part 4

### PR DESCRIPTION
**Pull request by yoomlam**
Connects #13130
Part 3 #13606

### Description
Now let's make some changes that will affect the task tree.
When creating a parent task, make sure it's in the same appeal.

### Acceptance Criteria
- [ ] All tests pass

### CODE REVIEW

File: spec/controllers/judge_assign_tasks_controller_spec.rb [hunk 1]
```diff
@@ -9,7 +9,7 @@
     let!(:judge_staff) { create(:staff, :judge_role, sdomainid: judge.css_id) }
     let!(:second_judge_staff) { create(:staff, :judge_role, sdomainid: second_judge.css_id) }
 
-    let!(:assign_tasks) { create_list(:ama_judge_task, 3, assigned_to: judge, parent: create(:root_task)) }
+    let!(:assign_tasks) { Array.new(3) { create(:ama_judge_task, assigned_to: judge, parent: create(:root_task)) } }
```
Comment (yoomlam) [inline spec/controllers/judge_assign_tasks_controller_spec.rb]: Intended behavior is 3 judge tasks, each with its own root task.

File: spec/jobs/hearing_disposition_change_job_spec.rb [hunk 2]
```diff
@@ -74,10 +74,10 @@ def create_disposition_task_for_legacy_hearings_ancestry(associated_hearing: nil
     # Property: Class that subclasses DispositionTask.
     context "when there are ChangeHearingDispositionTasks" do
       let!(:disposition_tasks) do
-        create_list(:assign_hearing_disposition_task, 6, parent: create(:hearing_task))
+        Array.new(6) { create(:assign_hearing_disposition_task, parent: create(:hearing_task)) }
       end
       let!(:change_disposition_tasks) do
-        create_list(:change_hearing_disposition_task, 3, parent: create(:hearing_task))
+        Array.new(3) { create(:change_hearing_disposition_task, parent: create(:hearing_task)) }
```
Comment (yoomlam) [inline spec/jobs/hearing_disposition_change_job_spec.rb]: Intended behavior is that each task has its own parent hearing task.

File: spec/factories/task.rb [hunk 3]
```diff
@@ -8,6 +8,19 @@
...
+          puts "Warning: Parent task #{task.parent&.id} is in different appeal than child task #{task.id}"
+        end
+
+        if task.parent&.appeal_type != task.appeal_type
+          puts "Warning: Parent task #{task.parent&.id} has different appeal_type #{task.parent&.appeal_type} " \
+               "vs. child task #{task.id} #{task.appeal_type}"
+        end
+      end
+    end
+
```
Comment (yoomlam) [inline spec/factories/task.rb]: Print warnings when the task tree is not realistic.

File: spec/controllers/tasks_controller_spec.rb [hunk 4]
```diff
@@ -185,7 +185,7 @@
           end
 
           get :index, params: { user_id: user.id, role: "unknown" }
-          expect(response).to have_http_status(:success)
+          expect(response).to be_successful
```
Comment (yoomlam) [inline spec/controllers/tasks_controller_spec.rb]: Addresses deprecation warning.

File: spec/factories/task.rb [hunk 5]
```diff
@@ -249,7 +262,7 @@
     end
 
     factory :ama_task, class: Task do
-      appeal
+      appeal { @overrides[:parent] ? @overrides[:parent].appeal : create(:appeal) }
```
Comment (yoomlam) [inline spec/factories/task.rb]: Handles the case where a parent task is specified. The new task should be in the same appeal as the parent.

File: spec/factories/task.rb [hunk 6]
```diff
@@ -264,6 +265,16 @@
...
+      factory :appeal_task, class: DecisionReviewTask do
+        appeal { create(:appeal, benefit_type: "education") }
+        assigned_by { nil }
+      end
+
+      factory :higher_level_review_task, class: DecisionReviewTask do
+        appeal { create(:higher_level_review, benefit_type: "education") }
+        assigned_by { nil }
+      end
+
```
Comment (yoomlam) [inline spec/factories/task.rb]: Pulled these in as subfactories under `ama_task`.

File: spec/controllers/judge_assign_tasks_controller_spec.rb (hunk 1 above)
Comment (lomky) [reply spec/controllers/judge_assign_tasks_controller_spec.rb]: What did the old way create?

Comment (lomky) [reply spec/controllers/judge_assign_tasks_controller_spec.rb]: ahh, three on the same appeal. interesting. I think there might be a more ruby-way to write this new syntax; I'mm come back to this!

File: spec/factories/task.rb (hunk 5 above)
Comment (yoomlam) [reply spec/factories/task.rb]: I use `@overrides` to check if a parent is specified when `create` is called. https://stackoverflow.com/questions/31205859/factory-girl-can-we-check-if-an-attribute-has-been-overridden-in-its-lazy-attri
> Though not documented (and therefore even more subject to change) you can access the overriden parameters in the instance variable @overrides.
This kind-of worries me, but it's a clean solution (i.e., one line of code).

File: spec/factories/task.rb (hunk 3 above)
Comment (yoomlam) [reply spec/factories/task.rb]: In Part 5, the `puts` become `fail`s.

File: spec/factories/task.rb (hunk 5 above)
Comment (lomky) [reply spec/factories/task.rb]: Is there a way we could implement it with a `transient` attribute?

See the appeal factory for an example:  https://github.com/department-of-veterans-affairs/caseflow/blob/master/spec/factories/appeal.rb#L45

Comment (yoomlam) [reply spec/factories/task.rb]: https://dsva.slack.com/archives/CJL810329/p1583790455235300?thread_ts=1583341605.139100&cid=CJL810329

Comment (yoomlam) [reply spec/factories/task.rb]: Added comment.

**Pull request merged**